File: 0452-minimum-number-of-arrows-to-burst-balloons/0452-minimum-number-of-arrows-to-burst-balloons.cpp
class Solution {
public:
    
    static bool cmp(vector<int>& a, vector<int>& b){
        return a[1] < b[1];
    }
    
    int findMinArrowShots(vector<vector<int>>& points) {
        
        if(points.size() < 2)
            return points.size();
        
        sort(points.begin(), points.end(), cmp);
        int arrows = 1;
        int prev = points[0][1];
        for(int i = 1;i < points.size(); ++i){
            if(points[i][0] > prev){
                prev = points[i][1];
                arrows++;
            }                
        }
        
        return arrows;
    }
};

File: 0452-minimum-number-of-arrows-to-burst-balloons/0452-minimum-number-of-arrows-to-burst-balloons.java
class Solution {
    public int findMinArrowShots(int[][] points) {
        
        int n = points.length;
        
        Arrays.sort(points, (a, b) -> {
            if(a[0] == b[0]) return b[1] - a[1];
            else return a[0] - b[0];
        });
        
        int count = 1;
        int si = points[0][0];
        int ei = points[0][1];
        for(int[] balloon : points){
            int start = balloon[0];
            int end = balloon[1];

            if(end < si || start > ei) {
                si = start;
                ei = end;
                count ++;
            } else {
                si = Math.max(si, start);
                ei = Math.min(ei, end);
            }            
        }

        return count;
    }
}